File: python/bestbefore.py
import itertools
import datetime

class BestBefore:

  def __init__(self):
    self.possible_dates = []
    
  def earliest_date(self, input):
    self.init_possible_dates(input) 
    if self.possible_dates:
      return min(self.possible_dates).strftime('%Y-%m-%d')
    else:
      return input + " is illegal"  
      
  def init_possible_dates(self, input):
    self.possible_dates = []
    date_parts = self.parse_input(input)
    for parts in map(list, itertools.permutations(date_parts)):
      try:
        if parts[0] < 2000:
          parts[0] += 2000
        self.possible_dates.append(datetime.date(parts[0], parts[1], parts[2]))
      except (ValueError, IndexError):
        pass
    
  def parse_input(self, input):
    date_parts = input.split('/')
    try:
      date_parts = map(int, date_parts)
    except ValueError:
      date_parts = []
    return date_parts

File: ruby/bestbefore.rb
require 'date'

class BestBefore

  def initialize
    @possible_dates = []
  end
  
  def earliest_date(input)
    init_possible_dates(parse_input(input)) 
    return "#{input} is illegal" if @possible_dates.empty?
    return "#{@possible_dates.min}" 
  end
  
  protected
 
  def init_possible_dates(date_parts) 
    @possible_dates = []
    date_parts.permutation.to_a.each do |parts|
      begin
        parts[0] += 2000 if parts[0] < 2000
        date = Date.new(parts[0], parts[1], parts[2])
        @possible_dates << date
      rescue NoMethodError, ArgumentError => e
      end
    end
  end
  
  def parse_input(input)
    input.split('/').map {|part| part.to_i }    
  end

end

